src/fsharp/FSharp.Core/async.fs

```diff
@@ -124,9 +119,8 @@ namespace Microsoft.FSharp.Control
                             action <- (fun () -> econt edi)
 
             finally
-                if thisIsTopTrampoline then
-                    Trampoline.thisThreadHasTrampoline <- false
-            fake()
+                Trampoline.thisThreadHasTrampoline <- not thisIsTopTrampoline
```

Comment:
You invert the value 'Trampoline.thisThreadHasTrampoline' twice.
once when saving it away in thisIsTopTrampoline 
and once when pushing it back to Trampoline.thisThreadHasTrampoline, it doesn't look like it used between those two spots.  Why don't you just store it away as was, and restore it as was?

Reply:
I replicated the previous logic without having seen that `thisIsTopTrampoline` was never used. It's fixed now.